### PR TITLE
[8.2/12] refactor(components): migrate mixed control+container (Tabs, DropdownMenu, Select) to role-named utilities

### LINKED ISSUE #124: [8/12] refactor(components): roll out role-named utilities to remaining 12 components
## Summary

Apply the role-named utility pattern (proven in #231 / `[7/12]` Button) to the remaining components. Architectural baseline established — `--control-h-*` was dropped in #229, the `cn()` merge-config was extended in #231, and the dimensional regression-sentinel pattern is in place.

**Status (post-#231 merge):**

- ✅ Coupling table — `.claude/rules/spacing-tokens.md` § Role-to-component coupling
- ✅ `cn()` extended for role-utility groups — covers `px-control-*`, `py-control-*`, `gap-*`, `p-container` (no changes needed per component in this sweep)
- ✅ Button migrated as proof point (#231)
- ⏳ `--control-gap-{sm,md,lg}` shape — #230 (open). **Should land before this sweep starts** so per-size gap is available from PR #1 instead of retrofitted across 6 control PRs.
- ⏳ Dimensional regression-sentinel helpers — #232 (open). Extracts `getControlHeight` / `expectModeCascadeWorks` / `expectHeightPinned` from `#231` Button stories. Lands the `toBe → toBeCloseTo`, dead-`await`, and 3-mode-chain fixes from `#231`'s review feedback so all 10 components inherit the corrected patterns.

## Prerequisites

- [ ] #230 lands (per-size `--control-gap-{sm,md,lg}`)
- [ ] #232 lands (dimensional regression-sentinel test-utils helpers)
- [ ] Each PR gets its own coupling-table audit (see Playbook step 1)

## Component scope reclassification

Re-reading the inventory against the post-#229 coupling table, the original 12 components split into 4 groups. Two are out of scope for spacing migration.

| Component    | Current utilities                               | Group                   | Notes                                                                                              |
| ------------ | ----------------------------------------------- | ----------------------- | -------------------------------------------------------------------------------------------------- |
| Tooltip      | `px-3 py-1.5`                                   | Pure control            | Mirrors Button — simplest migration                                                                |
| Input        | `px-{2.5,3,4} py-{1.5,2,3}`                     | Pure control            | Multi-size; mirrors Button's CVA structure                                                         |
| Badge        | `gap-1 p-0 px-2 px-2.5 py-0.5` + `size-{3.5,5}` | Control + icon variants | Decide square-padding variants like Button (icon) per the new spec note                            |
| Tabs         | `mt-2 p-1 px-{2,3,4} py-{1,1.5,2}`              | Control + container     | Trigger uses `px/py-control-*`; the list/root uses `p` and `mt`                                    |
| Accordion    | `pb-4 pt-0 py-4` + `size-4`                     | Internal-only           | `py` likely maps to `py-control-md` or `py-control-sm`; verify against item-tier role              |
| Alert        | `mb-1 p-4` + `size-4`                           | Container               | `p-4` likely → `p-container`; `mb-1` is sub-element offset (stays numeric)                         |
| Card         | `gap-{1.5,2} p-6 pt-0` + positional             | Container               | `p-6` → `p-container`; positional (`right-6`, `top-6`) stays numeric                               |
| Dialog       | `gap-{1.5,4} p-6` + positional                  | Container               | Container migration; **portal-into-scoped subtree caveat** (see Risks)                             |
| DropdownMenu | mixed padding/gap + `size-*`                    | Mixed                   | Container `p-1` outer + item-level `px/py`; same portal caveat                                     |
| Select       | similar to DropdownMenu                         | Mixed                   | Same portal caveat                                                                                 |
| Avatar       | `size-{5,6,8,10,12,14,16,20,24,full}`           | **OUT OF SCOPE**        | Coupling table line: "(uses sizing tokens, not spacing) — N/A — see `nx:size-*` numeric utilities" |
| Switch       | `h-5 w-9 size-4`                                | **OUT OF SCOPE**        | Track-and-thumb sizing, not control padding                                                        |

**Net components to migrate in this issue:** 10 (was 12 — Avatar / Switch dropped).

### Step 1 — Audit before edit

1. Read the component file end-to-end
2. Re-read the component's row in `spacing-tokens.md` § Role-to-component coupling
3. Math out vega/nova/sera intrinsic heights for each size variant. Padding-y × 2 + content (line-height or icon-size) should match designer intent.
4. Identify edge-case variants (icon-only, square padding, sub-pixel borders) and decide stays-numeric vs migrate vs add new role family. Add a row + footnote to `spacing-tokens.md` for any "stays-numeric" decision — Button (icon) is the template.

### Step 2 — Swap CVA class strings

- Replace numeric `nx:px-*`/`nx:py-*`/`nx:gap-*`/`nx:p-*` with role-named equivalents per the coupling table
- Keep numeric for: positional (`right-6`), sub-element offsets (`mb-1` inside Alert content), icon-sizing (`size-4`), and explicitly-numeric edge cases (e.g., Button's `nx:p-2.5`)
- DO NOT re-extend `cn()` — already covers role utilities globally
- Verify vega rendering byte-identical to pre-refactor (per coupling table values)

### Step 3 — AllModes story

- Import `SPACING_MODES` from `../../stories/spacing-modes`
- 7-row layout, each row scopes `data-style={mode}` on its container div
- Mode label: `nx:w-[64px] nx:typography-label-default nx:font-mono nx:text-muted-foreground` (mode-invariant width — `nx:w-16` would shift per mode)
- `parameters: { a11y: { test: 'off' } }` — duplicates of stories that already get a11y coverage
- Add a description noting that 4 of 7 modes (vega/lyra/luma/mira) currently share identical control tokens

### Step 4 — Dimensional regression sentinels (2 per migrated component)

Per the #231 review feedback (`toBe` flake-risk + mode-ordering coupling concerns):

```tsx
// Cascade sentinel — non-strict, robust to designer retunes
export const ModesProduceDifferentHeights: Story = {
  parameters: { a11y: { test: "off" } },
  render: () => (
    <div className="nx:flex nx:items-center nx:gap-4 nx:p-10 nx:bg-background">
      <div data-style="nova" data-testid="mode-host-nova">
        <Component />
      </div>
      <div data-style="sera" data-testid="mode-host-sera">
        <Component />
      </div>
    </div>
  ),
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const novaH = Math.round(
      canvas
        .getByTestId("mode-host-nova")
        .querySelector("button")!
        .getBoundingClientRect().height,
    );
    const seraH = Math.round(
      canvas
        .getByTestId("mode-host-sera")
        .querySelector("button")!
        .getBoundingClientRect().height,
    );
    expect(novaH).toBeLessThan(seraH); // 1 pair, not a 3-mode chain — less brittle
  },
};

// Contract pin — flake-safe via toBeCloseTo
export const VegaDefaultHeightPinned: Story = {
  parameters: { a11y: { test: "off" } },
  render: () => (
    <div
      data-style="vega"
      data-testid="vega-host"
      className="nx:p-10 nx:bg-background"
    >
      <Component />
    </div>
  ),
  play: async ({ canvasElement }) => {
    await document.fonts.ready;
    const canvas = within(canvasElement);
    const button = canvas.getByRole("button");
    expect(button.getBoundingClientRect().height).toBeCloseTo(EXPECTED_PX, 0);
  },
};
```

Update from #231 review feedback:

- `Math.round(...)` or `toBeCloseTo(N, 0)` — `toBe(N)` flakes on sub-pixel rendering
- Pair-wise comparison (`nova < sera`) — the original 3-mode chain (`nova < vega < sera`) hard-coded a design-ordering invariant; if a designer retunes Nova, the test fails for non-code reasons
- Drop `await` on synchronous matchers (`toBeLessThan` on a primitive is not a Promise)

### Step 5 — Verify

```bash
yarn workspace @nexus/react typecheck
yarn lint
yarn test:storybook
yarn workspace @nexus/react audit:storybook-coverage --component <name>
```

All four must be clean before merge.

## PR structure recommendation

12 separate PRs creates linear review burden. Grouping by family keeps each PR reviewable while bundling related architectural decisions.

**Proposed 3 batches:**

1. **Pure controls** — Tooltip, Input, Badge (3 components, ~450 LOC). Simplest; locks in the pattern for the rest of the sweep.
2. **Mixed control + container** — Tabs, DropdownMenu, Select (3 components, ~600 LOC). Adds the portal-caveat handling (see Risks below).
3. **Containers + internal** — Card, Dialog, Alert, Accordion (4 components, ~700 LOC).

Each batch is one PR. Each PR closes part of #124; final PR closes the issue.

Alternative: 10 separate PRs if the team prefers tighter bisects. Trade-off is the team's call.

### Portal interaction with `data-style` cascade

Dialog, DropdownMenu, Select, Tooltip use Radix portals — `<DialogContent>` renders at `document.body`, OUTSIDE any subtree `data-style` scope. So an `AllModes` story for Dialog will show:

- The trigger button varying per mode (it's inside the scope)
- The opened dialog rendering at the document-level default mode

Two options:

- Configure the Portal to render inside the scope (Radix `Portal container` prop)
- Accept the limitation, document in the story description

### Tabs roving focus

Tabs trigger uses Radix `RovingFocusGroup` — focus moves to the item under pointer hover. Per `components.md` § Surface exception map, "menu items" use background tint (not the outline ring). Verify Tabs trigger's focus pattern matches the rule before swapping spacing.

### Input's `aria-invalid` + focus

Currently `nx:aria-invalid:border-border-error nx:aria-invalid:focus-visible:outline-focus-error`. Spacing migration should not touch the focus chain, but the CVA's compound variants may reference padding indirectly — verify.

### Badge icon-only variants

If Badge supports icon-only (small dot or icon chips), it likely needs the same `nx:p-2.5`-stays-numeric decision as Button. Add a row to the coupling table per Step 1's audit.

### `nx:size-*` is not spacing

Avatar uses `nx:size-*`; Switch uses `nx:h-5 nx:w-9`. These are sizing tokens, not spacing role tokens. **They are not part of this migration.** If sizing tokens get their own role family later, that's a separate issue.

## Code-quality tripwires (from #231 review)

These came up in PR review; bake into PR #1 of this sweep so they don't repeat across all 12:

| Pattern                      | Avoid                                                             | Use                                                                 |
| ---------------------------- | ----------------------------------------------------------------- | ------------------------------------------------------------------- |
| Dimensional pin              | `expect(h).toBe(36)`                                              | `expect(h).toBeCloseTo(36, 0)` or `Math.round(h) === 36`            |
| Synchronous matcher          | `await expect(num).toBeLessThan(other)`                           | `expect(num).toBeLessThan(other)`                                   |
| Source comments              | JSDoc rationale block ("...because consumers would otherwise...") | One-liner describing what the code does; rationale lives in PR body |
| Test-coupled design ordering | `nova < vega < sera` (3-mode chain)                               | One pair (`nova < sera`) — robust to mid-mode retunes               |

## Acceptance criteria

- [ ] #230 has landed (per-size `--control-gap-{sm,md,lg}`)
- [ ] 10 components migrated (Tooltip, Input, Badge, Tabs, DropdownMenu, Select, Card, Dialog, Alert, Accordion)
- [ ] Per component: vega rendering byte-identical to pre-refactor
- [ ] Per component: `AllModes` story + 2 dimensional play tests
- [ ] Per component: `audit:storybook-coverage` clean
- [ ] All existing play tests pass
- [ ] Edge-case decisions (icon-only variants, sub-element offsets) documented in `spacing-tokens.md` coupling table at migration time
- [ ] Avatar and Switch confirmed out of scope (or moved to a sizing-token migration issue)

## Out of scope

- Migrating Avatar / Switch — sizing tokens, not spacing
- Tuning per-mode token values — separate concern (designer retunes)
- Lint rule rollout — `[11/12]` issue
- Component-level polish (animation, density alignment) — milestone #7 / epic #181

## References

- **Spec:** `.claude/rules/spacing-tokens.md` § Role-to-component coupling table
- **Proof point:** #231 (`[7/12]` Button migration)
- **Precursor:** #229 (drop `--control-h-*`)
- **Blocking dependency:** #230 (per-size `--control-gap-{sm,md,lg}`)
- **Milestone:** [Spacing tokens · Phase 1](https://github.com/nexuslabs-ai/nexus/milestone/5)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by aishvaryagupta**
## Summary

Batch 2 of 3 for the role-named utility rollout from #124. Migrates Tabs, DropdownMenu, and Select — the mixed control+container family. Vega rendering is byte-identical to pre-refactor for every migrated utility. The Input/Select cross-mode `px` drift design call (flagged in batch 1) is ratified here as **"accept the drift"** — see `spacing-tokens.md` update.

## Per-component changes

| Component | CVA migration | Sentinel pattern |
|-----------|---------------|------------------|
| **Tabs** | `default px-3 py-1.5` → `px-control-sm py-control-sm`; `lg px-4 py-2` → `px-control-md py-control-md`; `sm` stays numeric (sub-control rhythm); `TabsList p-1` + `TabsContent mt-2` stay (chrome) | Cascade `nova+maia` + pinned `vega=34px` |
| **DropdownMenu** | All 5 item-tier sub-components (`Item`, `CheckboxItem`, `RadioItem`, `SubTrigger`, `Label`) migrate `py-1.5` → `py-control-sm`. `px-2` stays (item-tier ≠ control-tier). Content/Sub-content `p-1`, Separator `-mx-1 my-1` stay (menu chrome). | Computed-style on opened item (portaled): `paddingTop === '6px'` in vega; close in `finally` |
| **Select** | Trigger `py-2 gap-2` → `py-control-md gap-control-md`; `px-3` stays (form-field convention). `Label` + `Item` migrate `py-1.5` → `py-control-sm` in lockstep with DropdownMenu. Scroll buttons + Separator stay. | Cascade `maia+sera` on trigger (not portaled, has role tokens) + pinned `vega=38px` |

## Coverage pair plan across the suite

Each component picks a different `(smaller, larger)` mode pair so the suite spans every distinct `control-*-y` value rather than testing the same one three times:

| Component | Pair | Values hit |
|-----------|------|------------|
| Button (#231) | nova+sera | control-md-y: 6, 12 |
| Input (#235) | nova+sera | control-md-y: 6, 12 |
| **Tabs (this PR)** | nova+maia | control-sm-y: 4, 8 |
| **Select (this PR)** | maia+sera | control-md-y: 10, 12 |

Between them the suite covers `control-sm-y` at 4/8 and `control-md-y` at 6/10/12 in vega-equivalent terms — every distinct value across the 7 modes.

## Design call recorded — "accept the drift"

Batch 1 left an open question: should Input/Select migrate `padding-x` to `control-x-md` (changes vega rendering for alignment with Button) or stay numeric (per-mode form-row gap drifts)? `spacing-tokens.md` now records the explicit decision: **accept the drift**. Per-mode form-row variance is treated as intentional design surface — compact modes render form fields tight against buttons, breathy modes give buttons visible chrome. Alternative paths (retune `--control-padding-x-md` to 12, or add `--input-padding-x-*` family) stay open if real complaints arrive.

A visual mockup of the drift across 4 distinct modes is at `reports/issue-124-gap-drift.html` (gitignored; was used to make the call).

## Council findings integrated

- **SDE2 caught DropdownMenu sub-components partial-migration** — fixed: all 5 item-tier sub-components migrate together (lockstep), not just `DropdownMenuItem`.
- **SDE2 caught SelectLabel** — fixed: included in lockstep migration.
- **Tester flagged portal teardown races** — DropdownMenu sentinel closes in `finally` via `{Escape}` so the open portal doesn't leak.
- **Tester flagged `toHaveClass` weakness for portaled content** — DropdownMenu sentinel uses `getComputedStyle().paddingTop` instead, which catches a typo'd role utility that `toHaveClass` would let through.

## GitHub Issue

Refs #124 (partial — closes on batch 3)

## Test Plan

- [x] `yarn typecheck` — clean
- [x] `yarn lint` — clean
- [x] `yarn test:storybook` — 265 tests pass (+8 from batch-1's 257), including 7 new mode/sentinel stories
- [x] `yarn audit:storybook-coverage --component tabs` — no gaps
- [x] `yarn audit:storybook-coverage --component dropdown-menu` — no gaps
- [x] `yarn audit:storybook-coverage --component select` — no gaps
- [x] `yarn format:check` — clean

## Next batch

Batch 3 — Card, Dialog, Alert, Accordion. Final PR closes #124.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .claude/rules/spacing-tokens.md [hunk 1]
```diff
@@ -216,12 +216,16 @@ When a new component is authored, the author adds a row to this table.
...
...e — Menu items (DropdownMenu, Select) keep `padding-x` numeric; only `padding-y` migrates.** Menu-item utilities are tighter than form controls — `DropdownMenuItem` / `DropdownMenuCheckboxItem` / `DropdownMenuRadioItem` / `DropdownMenuSubTrigger` / `DropdownMenuLabel` and `SelectItem` / `SelectLabel` all use `nx:px-2 nx:py-1.5` (8, 6). `py-1.5` matches `--control-padding-y-sm` (6px in vega) and migrates, giving menu rows the same vertical density-coupling as compact buttons. `px-2` (8px) sits below `--control-padding-x-sm` (12px) by design — a menu row is _denser_ than a control row, and forcing `px-control-sm` would visually merge adjacent rows into oversized buttons. Same item-tier rationale as Badge. The menu container chrome — `DropdownMenuContent` / `DropdownMenuSubContent` / `SelectContent` Viewport `nx:p-1` (4px), `DropdownMenuSeparator` / `SelectSeparator` `nx:-mx-1 nx:my-1`, `SelectScrollUpButton` / `SelectScrollDownButton` `nx:py-1` (4px) — stays numeric: these are sub-canonical-step offsets used to compose the popover frame, not container-tier padding. Cascade sentinels for these components run via computed-style on opened content (portaled), not dimensional measurement.
```
Comment (examlly-tech) [inline .claude/rules/spacing-tokens.md]: Coupling-table note covers menu-item `padding-x` staying numeric but is silent on `gap`. SelectTrigger migrates `gap-2`→`gap-control-md` while DropdownMenuItem keeps `gap-2`. Either extend the item-tier rationale to `gap` explicitly, or migrate menu-item `gap` to match — currently the coupling contract is ambiguous on gap for new menu-tier components.

File: packages/react/src/components/ui/Tabs.stories.tsx [hunk 2]
```diff
@@ -593,6 +599,118 @@ export const AllVariants: Story = {
...
+        <TabsList>
+          <TabsTrigger value="a">Default</TabsTrigger>
+        </TabsList>
+      </Tabs>
+    </div>
+  ),
+  play: async ({ canvasElement }) => {
+    await expectHeightPinned(within(canvasElement), 'vega-host', 34);
+  },
+};
```
Comment (examlly-tech) [inline packages/react/src/components/ui/Tabs.stories.tsx:713]: Tabs `sm` (`px-2 py-1`) is documented as intentionally mode-stable in spec, but unlike Badge there's no `expectHeightPinnedAcrossModes` sentinel locking it in. A future PR migrating `py-1`→`py-control-sm` would slip through. Consider adding a stability pin if `sm` is meant to stay sub-control.

**Review by examlly-tech — COMMENTED** [review]
## Principal Architect Review

**Verdict:** NEEDS DISCUSSION

### Architectural Concerns

| File:Line | Concern | Severity |
|-----------|---------|----------|
| `.claude/rules/spacing-tokens.md:227` | The menu-item coupling note covers `padding-x` staying numeric but is silent on `gap`. SelectTrigger migrates `gap-2`→`gap-control-md` (control-tier) while DropdownMenuItem keeps `gap-2` (item-tier). The item-tier rationale should extend to `gap` explicitly, or the divergence will resurface on future menu-tier components. | ⚠️ Minor |
| `Tabs.stories.tsx:712` | Tabs `sm` (`px-2 py-1`) is documented as intentionally mode-stable, but unlike Badge there is no `expectHeightPinnedAcrossModes` sentinel guarding it. A future migration of `py-1`→`py-control-*` would not fail a test. Consider adding a stability pin so spec intent is enforced, not just documented. | ⚠️ Minor |

Migration body is correct: byte-identity claims verified against `spacing-vega.json`; coupling-table contract holds; the design-call ratification in `spacing-tokens.md` is durable.

---
_Review performed against `.claude/rules/` guidelines_

**Review by examlly-tech — COMMENTED** [review]
## SDE2 Review

**Verdict:** APPROVED

No code-quality issues. Migration is mechanically sound:

- Byte-identity claims hold against `packages/core/tokens/semantic/spacing-vega.json` (`padding-x.sm=12 / .md=16`, `padding-y.sm=6 / .md=8`, `gap.md=8`).
- Sentinel patterns are right-shaped for each component: dimensional cascade + exact-px pin for non-portaled triggers (Tabs, SelectTrigger); `getComputedStyle().paddingTop` on portaled items (DropdownMenuItem) — correctly catches the typo case `toHaveClass` would miss, because a typo'd role utility doesn't emit a Tailwind class and `paddingTop` falls back to `0px` rather than `6px`.
- Portal cleanup via `try/finally` + `{Escape}` + `waitFor(...).toBeNull()` is correct; cleanup-masks-error is a known footgun of this pattern but acceptable for test code.
- `nx:` prefix rule compliance is clean across all migrated utilities.
- All three story files import shared helpers from `'../../stories/spacing-modes'` and `'../../stories/test-utils'` — no duplication.

---
_Review performed against `.claude/rules/` guidelines_

**Review by examlly-tech — APPROVED** [review]
## Principal Architect Follow-up — Verdict: APPROVED

### Unresolved Issues

_All previous issues resolved._

- Menu-item `gap` rationale extended in `.claude/rules/spacing-tokens.md` (item-tier contrast vs SelectTrigger control-tier).
- `TabsSmIsDensityStable` sentinel added in `Tabs.stories.tsx` (26px pin across nova/vega/sera; mirrors Badge).

---
_Review performed against `.claude/rules/` guidelines_

**Review by examlly-tech — APPROVED** [review]
## SDE2 Follow-up — Verdict: APPROVED

### Unresolved Issues

_All previous issues resolved._

- Doc claims (px-2 / py-1.5 / gap-2 on menu items; `gap-2`→`gap-control-md` on SelectTrigger) verified against `dropdown-menu.tsx` and `select.tsx`.
- New sentinel: import shape correct, expected px math sound (text-xs 16 + py-1×2 + border×2 = 26), selector `[data-slot="tabs-trigger"]` unambiguous per host.

---
_Review performed against `.claude/rules/` guidelines_
